src/ProductConstructionService/ProductConstructionService.Common/RedisMutex.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using Microsoft.Extensions.Logging;
+
+namespace ProductConstructionService.Common;
+
+public interface IRedisMutex
+{
+    public Task<T> EnterWhenAvailable<T>(string mutexName, Func<Task<T>> action, int lockTimeHours);
+}
+
+/// <summary>
+/// Mutex implemented using a redis cache meant for synchronization between replicas
+/// </summary>
+public class RedisMutex : IRedisMutex
+{
+    private readonly IRedisCacheFactory _cacheFactory;
+    private readonly ILogger<RedisMutex> _logger;
+
+    private const int MutexWakeUpTimeSeconds = 5;
+
+    public RedisMutex(IRedisCacheFactory cacheFactory, ILogger<RedisMutex> logger)
+    {
+        _cacheFactory = cacheFactory;
+        _logger = logger;
+    }
+
+    public async Task<T> EnterWhenAvailable<T>(
+        string mutexName,
+        Func<Task<T>> action,
+        int lockTimeHours = 1)
```

Comment:
Why not `TimeSpan`?

Reply:
Can't use it as a default parameter since it's not compile time const. It could be something like `time = default`, and then check `if (time == default)` but not sure if that's better

Reply:
I'd do that. Or I'd at least put it in minutes, not hours

Reply:
went with the `default` route